File: src/parser/factory.scala
package pg.parser

import pg.language.{Language}
import pg.parsingtable.{ParsingTable}
import pg.lexer.{Lexer}
import parser.{Parser}
import callback.{ASTConstructor, DefaultCallbackController}

object factory {
  /**
   * Parserを生成するためのファクトリクラス
   */
  object ParserFactory {
    def create(language: Language, parsing_table: ParsingTable): Parser = {
      val cc = new DefaultCallbackController(language)
      new Parser(new Lexer(language.lex,cc), language.grammar, parsing_table, cc)
    }
    def createAST(language: Language, parsing_table: ParsingTable): Parser = {
      val cc = new ASTConstructor(language)
      new Parser(new Lexer(language.lex,cc), language.grammar, parsing_table, cc)
    }
  }
}

File: src/parser/factory.ml
  (**
   * Parserを生成するためのファクトリクラス
   *)
module Make(D:sig
  val language: Language.language
  val parsingtable: Parsingtable.parsingTable
end) = struct

  module C = Callback.DefaultCallbackController.Make(struct let language=D.language end)
  module P = Parser.Make(C)
  module P2 = P.Make(struct let parsingtable=D.parsingtable end)
  include P2
end

module MakeAst(D:sig
  val language: Language.language
  val parsingtable: Parsingtable.parsingTable
end) = struct

  module C = Callback.ASTConstructor.Make(struct let language=D.language end)
  module P = Parser.Make(C)
  module P2 = P.Make(struct let parsingtable=D.parsingtable end)
  include P2
end
